### ISSUE TITLE
Assertion failure in cgc program (Word Completion) after adding cgc_allocate as an external api

### ISSUE THREAD

**Issue by PriyankaBose** [SVF-tools/SVF#165]
I am trying to run `wpa` on this [program](https://github.com/trailofbits/cb-multios/tree/master/challenges/WordCompletion).  However after adding this line `{"cgc_allocate", ExtAPI::EFT_A2R_NEW}` in `ExtAPI.cpp` to include `cgc_allocate()` to the list of ExtAPIs, I am getting the following assertion failure:

```
wpa: /lib/MemoryModel/PAG.cpp:306: NodeID PAG::getGepObjNode(NodeID, const LocationSet&): Assertion `false && "new gep obj node kind?"' failed.
Aborted (core dumped)
```

I have compiled this program with LLVM 7.0. The commit hash corresponding to the version of SVF I am using is 7a7647d98d6a3a437a82c786b237a3174a010490. The source of `cgc_allocate` can be found [here](https://github.com/trailofbits/cb-multios/blob/master/include/libcgc.c#L153).

**Comment by yuleisui** [comment]
Have you tried the lastest SVF version?

**Comment by PriyankaBose** [comment]
No, I haven't tried the latest one. We are using SVF for one of our projects and it is based on that commit.

**Comment by yuleisui** [comment]
Would you mind having a try? I think the problem should have been fixed in the new version.

**Comment by PriyankaBose** [comment]
The latest SVF solves the problem. Do you have an idea what might be causing the issue so that we can backport it to our local version?

**Comment by yuleisui** [comment]
It might be helpful to take a look at the relevant files of constraint nodes and pag nodes. The changes were very small from llvm 7.0.0 to llvm 9.0.0

**Comment by Machiry** [comment]
After further investigation, we found that one of the commits results in missing points-to, which could be the cause of our issue.

Specifically, See the commit: https://github.com/SVF-tools/SVF/commit/7c45e4dba4bbdb4da2859bcc3dbf1135d73e9605
Look at the changes made to the `PAGBuilder.cpp`: https://github.com/SVF-tools/SVF/commit/7c45e4dba4bbdb4da2859bcc3dbf1135d73e9605?diff=unified#diff-124284edcaad3363e9fb9ff5e852b068

Here:
```
 	NodeID vnArg = getValueNode(arg);
	NodeID dummy = pag->addDummyValNode();
        - NodeID obj = pag->addDummyObjNode();
           // This always returns 0  (because BlackHole enum value is 0)
           // and hence obj will always be 0      
	+  NodeID obj = pag->addBlackholeObjNode();
           // Thus, the following if condition will never be true and consequently we miss points to information
	  if (vnArg && dummy && obj) {
```

In short, the above commit **just avoids generating any points-to facts for external heap functions where the newly allocated object is returned indirectly through one of the arguments**.

I am not sure, I understand the rationale behind this commit and whether the issue for which this patch was made has been properly fixed.

Any help resolving this is greatly appreciated.
In any case, __the changes made by the commit: https://github.com/SVF-tools/SVF/commit/7c45e4dba4bbdb4da2859bcc3dbf1135d73e9605 to `PAGBuilder.cpp` should be reverted or modified__.

**Comment by yuleisui** [comment]
@Machiry Yes, your understanding is correct. I am overseas now and might not be very easy to look at the code now. If I remember correctly, returning a new object indirectly from an external library is difficult to model. If you revert the commit for line 753, would that simply pass your test case?

**Comment by Machiry** [comment]
No. It wont..we hit the initial assert :(
